Intermediate CTC + Stochastic depth

jaesong opened the pull request:
This implements [intermediate CTC](https://arxiv.org/abs/2102.03216) with [stochastic depth](https://arxiv.org/abs/1603.09382) for Transformer and Conformer encoders.

`stochastic_depth_rate` indicates the drop probability, similar to the dropout parameter. This might be confusing as the many papers use the notation of keep probability (`= 1 - drop_rate`). But I think it would be better to follow dropout notation.

The return type of Encoder needs to be changed due to intermediate CTC. To minimize the issue, the return type is changed only if `Encoder(intermediate_layers=...)` is given.

I would add some recipes soon.

sw005320 commented:
Thanks, @jaesong!
This is really cool!

sw005320 commented:
Can you fix some CI errors? (https://travis-ci.org/github/espnet/espnet/builds/773609667#L2756)

sw005320 reviewed (commented):
The code looks very good to me
- Can you add an integration test?
- Can you prepare a config file in some recipes? If it observes the improvement in that recipe, we can use it as a default config (by changing the symlink).

hirofumi0810 commented on espnet/nets/pytorch_backend/transformer/encoder.py:
```diff
@@ -298,9 +307,23 @@ def forward(self, xs, masks):
             xs = self.embed(xs)
-        xs, masks = self.encoders(xs, masks)
+
+        intermediate_outputs = []
+        for layer_idx, encoder_layer in enumerate(self.encoders):
+            xs, masks = encoder_layer(xs, masks)
+
+            if self.intermediate_layers is not None and layer_idx + 1 in self.intermediate_layers:
+                encoder_output = xs
+                # intermediate branches also require normalization.
+                if self.normalize_before:
+                    encoder_output = self.after_norm(encoder_output)
+                intermediate_outputs.append(encoder_output)
```
```suggestion
        if self.intermediate_layers is None:
            xs, masks = self.encoders(xs, masks)
        else:
            intermediate_outputs = []
            for layer_idx, encoder_layer in enumerate(self.encoders):
                xs, masks = encoder_layer(xs, masks)
    
                if layer_idx + 1 in self.intermediate_layers:
                    encoder_output = xs
                    # intermediate branches also require normalization.
                    if self.normalize_before:
                        encoder_output = self.after_norm(encoder_output)
                    intermediate_outputs.append(encoder_output)
```

hirofumi0810 commented on espnet/nets/pytorch_backend/conformer/encoder.py:
```diff
@@ -236,10 +245,25 @@ def forward(self, xs, masks):
             xs = self.embed(xs)
 
-        xs, masks = self.encoders(xs, masks)
+        intermediate_outputs = []
+        for layer_idx, encoder_layer in enumerate(self.encoders):
+            xs, masks = encoder_layer(xs, masks)
+
+            if self.intermediate_layers is not None and layer_idx + 1 in self.intermediate_layers:
+                # intermediate branches also require normalization.
+                encoder_output = xs
+                if isinstance(encoder_output, tuple):
+                    encoder_output = encoder_output[0]
+                    if self.normalize_before:
+                        encoder_output = self.after_norm(encoder_output)
+                intermediate_outputs.append(encoder_output)
+
```
```suggestion
        if self.intermediate_layers is None:
            xs, masks = self.encoders(xs, masks)
        else:
            intermediate_outputs = []
            for layer_idx, encoder_layer in enumerate(self.encoders):
                xs, masks = encoder_layer(xs, masks)
        
                if layer_idx + 1 in self.intermediate_layers:
                    # intermediate branches also require normalization.
                    encoder_output = xs
                    if isinstance(encoder_output, tuple):
                        encoder_output = encoder_output[0]
                        if self.normalize_before:
                            encoder_output = self.after_norm(encoder_output)
                    intermediate_outputs.append(encoder_output)

```

hirofumi0810 commented on espnet/nets/pytorch_backend/conformer/encoder_layer.py:
```diff
@@ -93,60 +96,69 @@ def forward(self, x_input, mask, cache=None):
+            skip_layer = torch.rand(1).item() < self.stochastic_depth_rate
+            stoch_layer_coeff = 1. / (1 - self.stochastic_depth_rate)
+
+        if not skip_layer:
```
```suggestion
        if skip_layer:
            if cache is not None:
                x = torch.cat([cache, x], dim=1)
            if pos_emb is not None:
                return (x, pos_emb), mask
            return x, mask
```

hirofumi0810 replied:
With this change, the below codes can be kept original.

hirofumi0810 commented on espnet/nets/pytorch_backend/transformer/encoder_layer.py:
```diff
@@ -68,34 +70,44 @@ def forward(self, x, mask, cache=None):
-
-        if cache is not None:
-            x = torch.cat([cache, x], dim=1)
+        skip_layer = False
```
Same as Conformer

hirofumi0810 commented on espnet/nets/pytorch_backend/transformer/encoder_layer.py:
```diff
@@ -41,6 +41,7 @@ def __init__(
         dropout_rate,
         normalize_before=True,
         concat_after=False,
+        stochastic_depth_rate=0.,
```
Can you update docstrings?

hirofumi0810 commented on espnet/nets/pytorch_backend/e2e_asr_transformer.py:
```diff
@@ -228,10 +246,14 @@ def forward(self, xs_pad, ilens, ys_pad):
             loss_ctc_data = None
         elif alpha == 1:
             self.loss = loss_ctc
+            if self.intermediate_ctc_weight > 0:
+                self.loss = (1 - self.intermediate_ctc_weight) * loss_ctc + self.intermediate_ctc_weight * loss_intermediate_ctc
```
Remove

hirofumi0810 commented on espnet/nets/pytorch_backend/e2e_asr_transformer.py:
```diff
@@ -213,6 +223,14 @@ def forward(self, xs_pad, ilens, ys_pad):
+                    loss_intermediate_ctc += loss_inter
+
+                loss_intermediate_ctc /= len(self.intermediate_ctc_layers)
+
```
```suggestion
            if self.intermediate_ctc_weight > 0:
                loss_ctc = (1 - self.intermediate_ctc_weight) * loss_ctc + self.intermediate_ctc_weight * loss_intermediate_ctc
```

jaesong replied:
This would alter the perceived loss curve (in log or tensorboard), as it becomes weighted sum of two losses, not the original CTC loss. Is it ok?
Usually the original CTC loss becomes smaller with intermediate CTC, but the weighted sum of two losses would become larger than before (as intermediate loss >= CTC loss).

hirofumi0810 commented on espnet/nets/pytorch_backend/e2e_asr_transformer.py:
```diff
@@ -228,10 +246,14 @@ def forward(self, xs_pad, ilens, ys_pad):
             loss_ctc_data = float(loss_ctc)
         else:
             self.loss = alpha * loss_ctc + (1 - alpha) * loss_att
+            if self.intermediate_ctc_weight > 0:
+                self.loss = (1 - alpha - self.intermediate_ctc_weight) * loss_att + alpha * loss_ctc + self.intermediate_ctc_weight * loss_intermediate_ctc
```
Remove

sw005320 commented:
@jaesong, could you fix the CI error https://github.com/espnet/espnet/pull/3274/checks?check_run_id=2753649675#step:8:13 and respond to our comments/request?

jaesong commented:
Updates after 2a95bf2:

- Addressed changes suggested by @hirofumi0810 (except for `loss_ctc` on discussion).
- Implemented intermediate CTC for Mask-CTC. (`loss_ctc` may need changes after discussion)
- Added sample recipe for WSJ.
- Fixed black and flake8 issues from CI. I'm not able to run pytest right now though.

sw005320 commented:
Some CI errors still exist https://github.com/espnet/espnet/pull/3274/checks?check_run_id=3010796351#step:8:42

sw005320 commented:
@jaesong, could you fix the CI error?
You can run some tests locally by following https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#4-unit-testing

jaesong commented:
Fixed the remaining lint error. Seems like CI failure is not related to this pull request. Not sure about coverage (it shows issues on some files I haven't changed)

sw005320 commented:
Can you add a test to pass the intermediate ctc loss in the following files?
https://github.com/espnet/espnet/blob/master/test/test_e2e_asr_conformer.py
https://github.com/espnet/espnet/blob/master/test/test_e2e_asr_maskctc.py
https://github.com/espnet/espnet/blob/master/test/test_e2e_asr_transformer.py
This will improve the coverage and make your implementation stable.

After that, I'll merge this PR.

jaesong commented:
All tests passed!

sw005320 commented:
Thanks!
This is a really good function.
We can also add this function to our main README.md

danpovey commented:
Did you guys see any WER improvements from either intermediate CTC or stochastic depth?

jaesong commented:
I'm the first author of intermediate CTC so I guess yes :)
In [our paper](https://arxiv.org/abs/2102.03216), we tried Transformer-CTC with various depth (from 6 layers to 48 layers). Intermediate CTC gives consistent improvement of WER. Stochastic Depth gives improvement if the layer is sufficiently deep (24 layers or more). Combination of both is better than either one.

Cescfangs commented:
> I'm the first author of intermediate CTC so I guess yes :)
> In [our paper](https://arxiv.org/abs/2102.03216), we tried Transformer-CTC with various depth (from 6 layers to 48 layers). Intermediate CTC gives consistent improvement of WER. Stochastic Depth gives improvement if the layer is sufficiently deep (24 layers or more). Combination of both is better than either one.

@jaesong, have you seen improvement under AR model? In our case(large dataset), there was about 5% relative improvement under NAR model, no improvement under AR model though.

D-X-Y commented:
@jaesong , thanks for your codes. Your implementation seems to be a soft-version of stochastic depth, `x <- x + 1 / (1 - p) * f(x)`. In the most vision applications, their implementation is the hard-version of stochastic depth, i.e., has the probability of`1-p` to use `x` and `p` to use `f(x)`. Have you tried the hard-version and how is the performance?
